Plugin homebrew

jcftang opened the pull request:
This a first pass at teaching bcfg2 about homebrew on OSX. Much of this is lifted from the MacPorts.py file.

AlexanderS commented:
You may want to drop the first 6 commits from this pull request, because they are already in master.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+        for pkg in pkgcache:
+            pkgname = pkg.split(' ')[0].strip()
+            version = pkg.split(' ')[1].strip()
+            self.logger.info(" pkgname: %s version: %s" % (pkgname, version))
```
This looks like it should be logged at a debug level, if at all.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+        self.installed = {}
+        for pkg in pkgcache:
+            pkgname = pkg.split(' ')[0].strip()
+            version = pkg.split(' ')[1].strip()
```
``pkgname, version = pkg.strip().split()``

Or, if version might contain spaces:

``pkgname, version = pkg.strip().split(None, 2)``

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+    def VerifyPackage(self, entry, _):
+        """Verify Package status for entry."""
+	self.logger.info("VerifyPackage: %s : %s" % (entry.get('name'),
+                                                     entry.get('version')))
```
Log this at debug.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+        """Verify Package status for entry."""
+	self.logger.info("VerifyPackage: %s : %s" % (entry.get('name'),
+                                                     entry.get('version')))
+        if 'version' not in entry.attrib:
```
This block of code is not reachable; ``Verify*`` is only called if ``canVerify`` succeeds, and ``canVerify`` checks that all attributes in ``self.__req__`` are present.

jcftang replied:
I'm not entirely sure what you suggest to do to resolve this particular issue. Shall I just remove this method?

stpierre replied:
No, just this ``if`` block.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+        """Remove extra packages."""
+	pkgline = ""
+	for pkg in packages:
+		pkgline += " " + pkg.get('name')
```
``pkgline = " ".join(p.get('name') for p in packages)``

stpierre replied:
Actually, see below.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+		pkgline += " " + pkg.get('name')
+
+        self.logger.info("Removing packages: %s" % pkgline)
+        self.cmd.run("/usr/local/bin/brew uninstall %s" % pkgline)
```
Use the list version of ``run()`` instead:

    pkg_names = [p.get('name') for p in packages]
    self.logger.info("Removing packages: %s" % pkg_names)
    self.cmd.run(["/usr/local/bin/brew", "uninstall"] + pkg_names)

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+
+	try:
+		self.logger.info("Running : brew install %s" % pkgline)
+		self.cmd.run("/usr/local/bin/brew install %s" % pkgline)
```
Use a list comprehension to build your list of package names, and the list version of ``run()`` here as well.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,82 @@
+	for pkg in packages:
+		pkgline += " " + pkg.get('name')
+
+	self.logger.info("About to install packages : " + pkgline)
```
You've already got a much better log message on line 78 below; this can be removed.

stpierre commented:
Aside from some fairly minor code issues, LGTM.

jcftang commented:
The comments sound reasonable, I'll fix up the patch and resubmit it.

jcftang commented:
I think I've addressed most of the comments from the above. However after rebasing to the master branch...

```
(bcfg2)jtang@awesome-o:bcfg2$ bcfg2 -C etc/bcfg2.conf  -v -q  -c tmp
Starting Bcfg2 client run at 1424109059.87
Failed to instantiate tool <class 'Bcfg2.Client.Tools.HomeBrew.HomeBrew'>
Traceback (most recent call last):
  File "/Users/jtang/develop/bcfg2/src/lib/Bcfg2/Client/__init__.py", line 458, in load_tools
    self.tools.append(tool(self.config))
TypeError: __init__() takes exactly 4 arguments (2 given)
Loaded tool drivers:
 Action   Bundle   POSIX    VCS      launchd
The following entries are not handled by any tool:
Package:homebrew:cowsay
```

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,74 @@
+    pkgtype = 'homebrew'
+    pkgtool = ('/usr/local/bin/brew install %s', ('%s', ['name']))
+
+    def __init__(self, logger, setup, config):
```
Looks like you may have based this file on the ``MacPorts`` driver from 1.3.x, not 1.4.x; lots of function signatures have changed. In particular, the ``__init__`` for client tools only takes a single argument now.

jcftang replied:
Yes, that seems to be the case.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,74 @@
+        """Remove extra packages."""
+        pkg_names = " ".join(p.get('name') for p in packages)
+        self.logger.info("Removing packages: %s" % pkg_names)
+        self.cmd.run(["/usr/local/bin/brew", "uninstall"] + pkg_names)
```
``pkg_names`` shouldn't be a string; this addition will fail, and also it'll take the entire list of names as a single argument, which is not what you want. It should be a list, e.g.:

``pkg_names = [p.get('name') for p in packages]``

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,74 @@
+        self.RefreshPackages()
+        self.extra = self.FindExtra()
+
+    def InstallPackage(self, packages, states):
```
``Install*`` no longer takes a second argument in 1.4.x.

stpierre replied:
Actually, ``PkgTool`` doesn't even use ``InstallPackage``; it automatically generates the install command from ``cls.pkgtool``. This entire method can be removed.

jcftang commented:
The proposed changes have been made

jcftang commented:
Oops, I need to cleanup the last commit

solj commented:
Can you also add this to the 1.4.0pre2 release notes?

solj commented:
:+1:

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,58 @@
+
+    def VerifyPackage(self, entry, _):
+        """Verify Package status for entry."""
+        self.logger.info("VerifyPackage: %s : %s" % (entry.get('name'),
```
Log this at debug.

stpierre commented on src/lib/Bcfg2/Client/Tools/HomeBrew.py:
```diff
@@ -0,0 +1,58 @@
+        if 'version' not in entry.attrib:
+            self.logger.info("Cannot verify unversioned package %s" %
+                             entry.attrib['name'])
+            return False
```
This if block is unreachable.

stpierre commented:
Github seems to have eaten a couple of my older comments -- maybe it got confused in the rebase.
